[feat] Move QueryBuilder from sdk-core to circuit

ytham opened the pull request:
- Moves QueryBuilder from sdk-core to sdk-client/circuit
- Removes `AxiomSdkCore` (QueryBuilder is instantiated directly)
- Streamlines QueryBuilder
- Updates all of the tests
- Removes unused subquery builders

jonathanpwang commented on circuit/src/queryBuilder/configure.ts:
```diff
@@ -0,0 +1,73 @@
+
+export function handleChainId(config: AxiomV2QueryBuilderConfig): AxiomV2QueryBuilderConfig {
+  if (config.chainId === undefined) {
+    config.chainId = 1;
```
does `chainId` refer to `sourceChainId`?

jonathanpwang commented on circuit/src/queryBuilder/types.ts:
```diff
@@ -0,0 +1,82 @@
+  AxiomV2FieldConstant,
+} from "@axiom-crypto/tools";
+
+export interface AxiomV2QueryBuilderConfig {
```
Add dochints, especially on distinction between `chainId` and `targetChainId`.

jonathanpwang commented on circuit/src/queryBuilder/types.ts:
```diff
@@ -0,0 +1,82 @@
+  chainId?: number | string | BigInt;
+  targetChainId?: number | string | BigInt;
+  version?: string;
+  privateKey?: string;
```
remove `privateKey`: this should not be owned by QueryBuilder

jonathanpwang commented on circuit/src/queryBuilder/types.ts:
```diff
@@ -0,0 +1,82 @@
+  mock: boolean;
+  provider: ethers.JsonRpcProvider;
+  version: string;
+  signer?: ethers.Wallet;
```
remove `signer`: the `circuit` package should not own anything related to signer

jonathanpwang commented on circuit/src/queryBuilder/dataSubquery/configLimitManager.ts:
```diff
@@ -0,0 +1,177 @@
+import {
```
Assuming `dataSubquery/` is all copied from `sdk-core`, did not carefully review

jonathanpwang commented on circuit/src/queryBuilder/queryBuilder.ts:
```diff
@@ -0,0 +1,614 @@
+export class AxiomV2QueryBuilder {
+  readonly config: InternalConfig;
+  private builtQuery?: BuiltQueryV2;
+  private builtDataQuery?: AxiomV2DataQuery;
```
when do you have `builtDataQuery` defined but `builtQuery` not? can we just only have `builtQuery`?

ytham replied:
this was previously added to have a way to have circuit specify a dataQuery directly, but with the removal of `UnbuiltSubquery` types I was able to refactor this out

jonathanpwang commented on circuit/src/queryBuilder/queryBuilder.ts:
```diff
@@ -0,0 +1,614 @@
+
+export class AxiomV2QueryBuilder {
+  readonly config: InternalConfig;
+  private builtQuery?: BuiltQueryV2;
```
add some doc hints just so it's easier to keep track of the different roles of these prviate variables, especially around what needs to be provided vs built

jonathanpwang commented on circuit/src/queryBuilder/types.ts:
```diff
@@ -0,0 +1,82 @@
+  dataQueryStruct: AxiomV2DataQuery;
+  computeQuery: AxiomV2ComputeQuery;
+  querySchema: string;
+  callback: AxiomV2Callback;
```
`callback, userSalt, feeData, refundee` should not be owned by `BuiltQueryV2`: these are not `circuit` concepts - they are contract concepts, which should be owned by `client/`.

jonathanpwang replied:
anything involving these types or options should be moved to `client`. `circuit` should only have things that pertain to the data query or compute query itself.

jonathanpwang commented on circuit/src/queryBuilder/types.ts:
```diff
@@ -0,0 +1,82 @@
+  signer?: ethers.Wallet;
+}
+
+export interface AxiomV2QueryOptions {
```
This should be in `client/`, not `circuit/`. See comment below.

jonathanpwang commented on circuit/src/scaffold.ts:
```diff
@@ -23,7 +20,7 @@ export abstract class AxiomBaseCircuitScaffold<T> extends BaseCircuitScaffold {
   protected provider: string;
   protected dataQuery: DataSubquery[];
-  protected axiom: AxiomSdkCore;
+  protected axiom: AxiomV2QueryBuilder;
```
if this is purely internal, rename from `axiom` to just `qb` or something

jonathanpwang commented on circuit/src/scaffold.ts:
```diff
@@ -23,7 +20,7 @@ export abstract class AxiomBaseCircuitScaffold<T> extends BaseCircuitScaffold {
   protected dataQuery: DataSubquery[];
-  protected axiom: AxiomSdkCore;
+  protected axiom: AxiomV2QueryBuilder;
   protected computeQuery: AxiomV2ComputeQuery | undefined;
```
there is already a computeQuery in queryBuilder: redundancy? can you just use the one in queryBuilder. Otherwise it will be error-prone.

jonathanpwang commented on circuit/src/scaffold.ts:
```diff
@@ -23,7 +20,7 @@ export abstract class AxiomBaseCircuitScaffold<T> extends BaseCircuitScaffold {
   protected halo2Lib!: Halo2LibWasm;
   protected provider: string;
   protected dataQuery: DataSubquery[];
```
there is already dataQuery in queryBuilder: can it be re-used? otherwise the redundancy is going to be error prone.

ytham replied:
was able to remove queryBuilderBase from here

jonathanpwang commented on circuit/src/scaffold.ts:
```diff
@@ -23,7 +20,7 @@ export abstract class AxiomBaseCircuitScaffold<T> extends BaseCircuitScaffold {
   protected halo2Lib!: Halo2LibWasm;
   protected provider: string;
```
queryBuilder already has provider, redundancy.

jonathanpwang commented on circuit/src/scaffold.ts:
```diff
@@ -23,7 +20,7 @@ export abstract class AxiomBaseCircuitScaffold<T> extends BaseCircuitScaffold {
   protected halo2Lib!: Halo2LibWasm;
```
resultLen is already in computeQuery which is in both queryBuilder and this.computeQuery. I guess there is some ordering about when each one becomes defined. but if possible, try to remove redundancy.

ytham replied:
removed queryBuilderBase; will keep this.resultLen because it is used to set the value this.computeQuery

jonathanpwang commented on circuit/src/scaffold.ts:
```diff
@@ -23,7 +20,7 @@ export abstract class AxiomBaseCircuitScaffold<T> extends BaseCircuitScaffold {
-  protected axiom: AxiomSdkCore;
+  protected axiom: AxiomV2QueryBuilder;
   protected computeQuery: AxiomV2ComputeQuery | undefined;
   protected chainId: string;
```
already in queryBuilder

jonathanpwang commented on circuit/src/scaffold.ts:
```diff
@@ -224,15 +221,14 @@ export abstract class AxiomBaseCircuitScaffold<T> extends BaseCircuitScaffold {
     // Validate data subqueries
-    const queryBuilder = (this.axiom.query as QueryV2).new();
-
+    // const queryBuilder = (this.axiom.query as QueryV2).new();
```
delete?

jonathanpwang commented on circuit/src/scaffold.ts:
```diff
@@ -154,8 +151,8 @@ export abstract class AxiomBaseCircuitScaffold<T> extends BaseCircuitScaffold {
     // Validate max circuit subquery size
-    const queryBuilder = (this.axiom.query as QueryV2).new();
-    queryBuilder.setBuiltDataQuery({
+    // const queryBuilder = (this.axiom.query as QueryV2).new();
```
delete?

jonathanpwang commented on client/src/axiom/utils.ts:
```diff
@@ -67,15 +70,15 @@ export function argsObjToArr(
   let contractMinMaxFeePerGas = await publicClient.readContract({
-    address: axiom.getAxiomQueryAddress() as `0x${string}`,
-    abi: axiom.getAxiomQueryAbi(),
+    address: getAxiomV2QueryAddress(axiom.config.chainId.toString()) as `0x${string}`,
```
how does this play with the overrides?

jonathanpwang reviewed (changes requested):
This will streamline things a lot!
Requesting changes, two main categories:
- `circuit` should not own anything involving the contract, such as fees, callback, gas options. Those should be owned by `client`.
- circuit scaffold seems to still have a bunch of redundancy with `QueryBuilder`. We should look through those and remove as much redundancy as possible, since they can lead to errors later.

ytham commented:
@jonathanpwang this is ready for review again.

Also manually tested on base w/ setting the following options:
```
    options: {
      maxFeePerGas: "520000000",
      callbackGasLimit: 200000,
      overrideAxiomQueryFee: "30000000000000000",
      refundee: "0x3b49DE82B86d677C072Dcc7ED47bcA9F20f0CF46",
    }
```
which matches the logged values: https://sepolia.basescan.org/tx/0x89e68f1e024f70ec1f083855eabf41bf6bf4fc3668d71fb70b0c0f83e789a14f#eventlog

ytham commented:
not sure why these unit tests are now failing in CI; they are working locally. I've updated the github actions secrets just to make sure we're not using an old and invalid value for `PROVIDER_URI_MAINNET`. will dig into it more tomorrow...

ytham commented:
@jonathanpwang ready for review. split out QueryBuilder into QueryBuilderBase and QueryBuilderClient and separated concerns across the layers.

jonathanpwang commented:
Merge conflict?

jonathanpwang commented:
Can you mark my old comments as resolved if they've been resolved?

ytham commented on circuit/src/pkg/index.ts:
keeps the @axiom-crypto/circuit imports a bit cleaner. I've just refactored it slightly

jonathanpwang commented on circuit/src/queryBuilderBase/queryBuilderBase.ts:
```diff
@@ -0,0 +1,465 @@
+        this.dataSubqueryCount.account++;
+        if (this.dataSubqueryCount.account > ConstantsV2.MaxSameSubqueryType) {
+          throw new Error(
+            `Cannot add more than ${ConstantsV2.MaxSameSubqueryType} Account + Storage + Nested Mapping subqueries`,
```
is this message outdated?

ytham replied:
do we want to keep the same subquery type check or just check total number since the values are the same now?

ytham commented on harness/test/integration/start.test.ts:
```diff
@@ -55,46 +55,46 @@ describe("Integration tests", () => {
-    }
-    await expect(testFn()).rejects.toThrowError();
-  });
+  // test(`Size 129 header`, async () => {
```
ah was for testing but by mistake

ytham commented:
> It is very confusing to me why there is `AxiomCircuit` and `QueryBuilderClient` and they are both inside `Axiom` class. I would expect that ideally there is a single `Axiom` class that handles everything. If `QueryBuilderClient` is only constructed internally on-demand, that is also okay.
> 
> Is the `js/` vs `web/` divide for `AxiomCircuit` still necessary? If not, we don't need `AxiomCircuit` anymore.
> 
> I also didn't see any unit tests of the `Axiom` class itself - is that right?

i'll see what i can do to refactor. 

we unfortunately still need js/web since `Axiom` class uses js and the react webworker uses `web`.

currently no unit tests in client but the harness tests cover most of the functionality. however, i'll go ahead and add some unit tests to for additional coverage.

ytham commented:
@jonathanpwang I went ahead and made a number of modifications:
- QueryBuilderClient is no longer passed around and is only instantiated within functions
- added unit tests for `Axiom` class
- removed per-type subquery check and unified subquery checks to just look at total number

jonathanpwang reviewed (approved):
LGTM
